src/compiler/array.rs

```diff
@@ -1005,3 +1005,143 @@ pub(crate) fn slice(
 
     Ok(new_arr)
 }
+
+fn copy_from_element(
+    compiler: &mut Compiler,
+    source_element: &Symbol,
+    target_data_ptr: &Symbol,
+    target_capacity: &Symbol,
+    element_width: u32,
+    index: u32,
+) -> Result<()> {
+    // Ensure that the target array has enough capacity to hold the source element
+    compiler.instructions.extend([
+        Instruction::MemLoad(Some(target_capacity.memory_addr)),
+        // [capacity]
+        Instruction::Push(index + 1),
+        // [index + 1, capacity]
+        Instruction::U32CheckedGTE,
+        // [capacity >= index + 1]
+        Instruction::Assert,
+    ]);
+
+    // Copy the source element to the target array at the specified position
+    compiler.memory.read(
+        compiler.instructions,
+        target_data_ptr.memory_addr,
+        target_data_ptr.type_.miden_width(),
+    );
+    // [target_data_ptr]
+    for i in 0..element_width {
+        let offset = index * element_width + i;
+        compiler.instructions.extend([
+            // [target_data_ptr]
+            Instruction::MemLoad(Some(source_element.memory_addr + i)),
+            // [value, target_data_ptr]
+            Instruction::Dup(Some(1)),
+            // [target_data_ptr, value, target_data_ptr]
+            Instruction::Push(offset),
+            // [offset, target_data_ptr, value, target_data_ptr]
+            Instruction::U32CheckedAdd,
+            // [ptr = target_data_ptr + offset, value, target_data_ptr]
+            Instruction::MemStore(None),
+            // [target_data_ptr]
+        ]);
+    }
+    compiler.instructions.push(Instruction::Drop);
+    // []
+
+    Ok(())
+}
+
+pub(crate) fn unshift(
+    compiler: &mut Compiler,
+    arr: &Symbol,
+    elements: &[Symbol],
+) -> Result<Symbol> {
+    let element_type = element_type(&arr.type_);
+    for el in elements {
+        ensure_eq_type!(el, @element_type);
+    }
+
+    compiler
+        .memory
+        .read(compiler.instructions, array::length(arr).memory_addr, 1);
+    // [len]
+    compiler
+        .instructions
+        .push(encoder::Instruction::Push(elements.len() as u32));
+    // [elements.len(), len]
+    compiler
+        .instructions
+        .push(encoder::Instruction::U32CheckedAdd);
+    // [len + elements.len()]
+    let new_len = compiler
+        .memory
+        .allocate_symbol(Type::PrimitiveType(PrimitiveType::UInt32));
+    compiler.memory.write(
+        compiler.instructions,
+        new_len.memory_addr,
+        &[ValueSource::Stack],
+    );
+    // []
+
+    let new_arr = dynamic_new(compiler, element_type.clone(), new_len.clone())?;
+    for i in 0..elements.len() {
```

Comment:
as an option:

```suggestion
    for (i, el) in elements.iter().enumerate() {
```

then `elements[i]` -> `el`